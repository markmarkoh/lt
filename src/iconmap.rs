// linear to nerdfont
pub fn ico_to_nf(name: &str) -> String {
    let nf = match name {
        "Umbrella" => "  ",
        "FaceStarEyes" => "  ",
        _ => "idk"
    };
    nf.to_string()
}

pub fn p_to_nf(priority: f64) -> String {
    let nf = match priority {
        1.0 => "",
        2.0 => "󰕾",
        3.0 => "󰖀",
        4.0 => "󰕿",
        _ => "󰸈",
    };
    nf.to_string()
}
