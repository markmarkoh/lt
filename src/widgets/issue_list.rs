use std::sync::{Arc, RwLock};

use crossterm::event::{Event, KeyCode, KeyEventKind};
use ratatui::{
    buffer::Buffer,
    layout::Rect,
    style::{
        Modifier, Style, Stylize,
        palette::{
            material::{AMBER, BLUE_GRAY},
            tailwind::SLATE,
        },
    },
    text::{Line, Span, Text},
    widgets::{Block, List, ListItem, ListState, Paragraph, StatefulWidget, Widget, Wrap},
};

use crate::{
    LTWidget, LoadingState, LtEvent, IssueFragment,
    api::LinearClient,
    iconmap,
    queries::{
        MyIssuesQuery,
        my_issues_query::{self},
    },
};

#[derive(Debug, Default)]
pub struct MyIssuesWidgetState {
    loading_state: LoadingState,
    pub list_state: ListState,
    pub issues: Vec<IssueFragment>,
}

#[derive(Debug, Clone, Default)]
pub struct MyIssuesWidget {
    pub state: Arc<RwLock<MyIssuesWidgetState>>,
}

impl MyIssuesWidget {
    async fn fetch(self) {
        self.set_loading_state(LoadingState::Loading);
        let linear_api_token =
            std::env::var("LINEAR_API_TOKEN").expect("Missing LINEAR_API_TOKEN env var");
        let client = LinearClient::new(linear_api_token).unwrap();
        let variables = my_issues_query::Variables {};
        match client.query(MyIssuesQuery, variables).await {
            Ok(data) => {
                self.state.write().unwrap().issues = data.issues.nodes.iter().map(|issue| issue.to_owned().into()).collect();
                // TODO: Add cacheing implementation
                //println!("{:#?}", serde_json::to_string(&self.state.read().unwrap().issues).unwrap());
            }
            Err(e) => {
                self.set_loading_state(LoadingState::Error(e.to_string()));
                return;
            }
        }
        self.set_loading_state(LoadingState::Loaded);
    }

    fn set_loading_state(&self, state: LoadingState) {
        self.state.write().unwrap().loading_state = state;
    }

    fn get_loading_state(&self) -> LoadingState {
        self.state.read().unwrap().loading_state.clone()
    }

    pub fn scroll_down(&self) {
        let mut state = self.state.write().unwrap();
        if let Some(index) = state.list_state.selected() {
            if index >= state.issues.len() - 1 {
                return state.list_state.select_first();
            }
        }
        state.list_state.select_next()
    }

    pub fn scroll_up(&self) {
        let mut state = self.state.write().unwrap();

        match state.list_state.selected() {
            Some(0) | None => {
                let max_index = state.issues.len() - 1;
                state.list_state.select(Some(max_index));
            }
            _ => state.list_state.select_previous(),
        }
    }

    pub fn copy_branch_name(&self) {
        let state = self.state.read().unwrap();
        if let Some(index) = state.list_state.selected() {
            let selected_issue = &state.issues[index];
            cli_clipboard::set_contents(selected_issue.branch_name.clone()).unwrap();
        }
    }

    pub fn open_url(&self) -> std::result::Result<(), std::io::Error> {
        let state = self.state.read().unwrap();
        if let Some(index) = state.list_state.selected() {
            let selected_issue = &state.issues[index];
            open::that(&selected_issue.url)
        } else {
            Ok(())
        }
    }
}

impl LTWidget for MyIssuesWidget {
    fn run(&self) {
        let this = self.clone();
        tokio::spawn(this.fetch());
    }

    fn handle_event(&self, event: &Event) -> LtEvent {
        if self.get_loading_state() != LoadingState::Loaded {
            return LtEvent::None;
        }
        if let Event::Key(key) = event {
            if key.kind == KeyEventKind::Press {
                match key.code {
                    KeyCode::Char('j') => {
                        self.scroll_down();
                        return LtEvent::SelectIssue;
                    }
                    KeyCode::Char('k') => {
                        self.scroll_up();
                        return LtEvent::SelectIssue;
                    }
                    KeyCode::Char('r') => {
                        self.run();
                        // TODO: Figure out how to get state to update better
                        return LtEvent::SelectIssue;
                    }
                    KeyCode::Char('o') => {
                        let _ = self.open_url();
                        return LtEvent::None;
                    }
                    KeyCode::Char('c') | KeyCode::Char('y') => {
                        self.copy_branch_name();
                        return LtEvent::None;
                    }
                    _ => {
                        return LtEvent::None;
                    }
                };
            }
        }
        LtEvent::None
    }
}
const SELECTED_STYLE: Style = Style::new().bg(SLATE.c100).fg(BLUE_GRAY.c900);

impl Widget for &MyIssuesWidget {
    fn render(self, area: Rect, buf: &mut Buffer) {
        let mut block = Block::bordered()
            .title("My Issues")
            .title_bottom(Line::from(vec![
                Span::from(" <j/k> ").blue(),
                Span::from("to select issue "),
            ]));

        if let LoadingState::Loading = self.get_loading_state() {
            block = block.title(Line::from("Loadingâ€¦").right_aligned());
        }

        if let LoadingState::Error(e) = self.get_loading_state() {
            let p = Paragraph::new(vec![
                Line::from("Error:\n\n".red().bold()),
                Line::from(e.red().bold().underlined()),
            ])
            .wrap(Wrap { trim: true })
            .block(block);
            p.render(area, buf);
            return;
        }
        let mut state = self.state.write().unwrap();
        let area_width = area.width;
        let rows = state.issues.iter().map(|item| {
            let mut text = Text::default();
            let priority_icon = iconmap::p_to_nf(item.priority);
            let status_icon = iconmap::state_to_nf(&item.state.type_);
            let identifier = item.identifier.clone();
            // gives the effect of right aligning icons and left aligning the text
            let spaces = (area_width as usize) - identifier.len() - 8;
            let line = format!(
                "{}{}{}  {}",
                identifier.fg(AMBER.c700),
                " ".repeat(spaces),
                status_icon,
                priority_icon
            );
            text.extend([
                item.title.clone().white(),
                line.add_modifier(Modifier::BOLD).blue(),
            ]);
            ListItem::new(text)
        });

        // tests can't see the highlighting
        let highlight_symbol = if cfg!(test) {
            ">"
        } else {
            ""
        };

        let list = List::new(rows).highlight_style(SELECTED_STYLE).highlight_symbol(highlight_symbol).block(block);
        StatefulWidget::render(list, area, buf, &mut state.list_state);
    }
}

#[cfg(test)]
mod tests {
    use std::sync::{Arc, RwLock};

    use crossterm::event::{KeyCode, KeyEventKind, KeyEventState, KeyModifiers};
    use insta::assert_snapshot;
    use ratatui::{Terminal, backend::TestBackend, widgets::ListState};

    use crate::{
        LTWidget, LtEvent, queries,
        widgets::{self, MyIssuesWidget, selected_issue::tests::make_issue},
    };

    fn create_key_event(key: char) -> crossterm::event::Event {
        crossterm::event::Event::Key(crossterm::event::KeyEvent {
            code: KeyCode::Char(key),
            kind: KeyEventKind::Press,
            modifiers: KeyModifiers::empty(),
            state: KeyEventState::empty(),
        })
    }

    #[test]
    fn test_empty_state() {
        let app = MyIssuesWidget::default();
        let mut terminal = Terminal::new(TestBackend::new(40, 20)).unwrap();
        terminal
            .draw(|frame| frame.render_widget(&app, frame.area()))
            .unwrap();

        assert_snapshot!(terminal.backend());

        let ev = app.handle_event(&create_key_event('j'));
        assert_eq!(ev, LtEvent::None);
    }

    #[test]
    fn test_with_issues() {
        let issues = vec![
            make_issue("Ticket One", "TEST-1"),
            make_issue("Ticket Two", "TEST-2"),
        ];
        let app = MyIssuesWidget {
            state: Arc::new(RwLock::new(widgets::issue_list::MyIssuesWidgetState {
                loading_state: crate::LoadingState::Loaded,
                list_state: ListState::default(),
                issues
            })),
        };
        let mut terminal = Terminal::new(TestBackend::new(40, 20)).unwrap();
        terminal
            .draw(|frame| frame.render_widget(&app, frame.area()))
            .unwrap();

        assert_snapshot!(terminal.backend());

        assert_eq!(app.state.read().unwrap().list_state.selected(), None);

        let ev = app.handle_event(&create_key_event('j'));
        assert_eq!(ev, LtEvent::SelectIssue);
        assert_eq!(app.state.read().unwrap().list_state.selected(), Some(0));

        app.handle_event(&create_key_event('j'));
        assert_eq!(app.state.read().unwrap().list_state.selected(), Some(1));

        terminal
            .draw(|frame| frame.render_widget(&app, frame.area()))
            .unwrap();

        assert_snapshot!(terminal.backend());
        // test that is passes through back to 0
        app.handle_event(&create_key_event('j'));
        assert_eq!(app.state.read().unwrap().list_state.selected(), Some(0));

        // test that is passes backwards to 1
        app.handle_event(&create_key_event('k'));
        assert_eq!(app.state.read().unwrap().list_state.selected(), Some(1));

        // test that <y> yanks the branch name to the clipboard
        app.handle_event(&create_key_event('y'));
        assert_eq!(cli_clipboard::get_contents().unwrap(), "test-1-branch-name");

    }
}
