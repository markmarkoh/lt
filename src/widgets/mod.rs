mod selected_issue;
mod issue_list;
mod tab_widget;

pub use selected_issue::SelectedIssueWidget;
pub use issue_list::MyIssuesWidget;
pub use tab_widget::TabWidget;
