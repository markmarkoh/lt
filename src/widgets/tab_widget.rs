use crossterm::event::{Event, KeyCode, KeyEventKind};
use ratatui::buffer::Buffer;
use ratatui::layout::{Constraint, Layout};
use ratatui::style::Color;
use ratatui::text::{Line, Span};
use ratatui::{
    layout::Rect,
    style::Stylize,
    widgets::{Tabs, Widget},
};
use std::str::FromStr;
use std::sync::{Arc, RwLock};

use crate::api::LinearClient;
use crate::queries::{CustomViewsQuery, custom_views_query};
use crate::{TabChangeEvent, iconmap};

#[derive(Debug, Clone)]
pub struct TabWidget {
    state: Arc<RwLock<TabWidgetState>>,
}

#[derive(Debug, Clone)]
struct TabWidgetState {
    selected_index: usize,
    tabs: Vec<Tab>,
}

impl Default for TabWidget {
    fn default() -> Self {
        TabWidget {
            state: Arc::new(RwLock::new(TabWidgetState {
                selected_index: 0,
                tabs: vec![Tab {
                    title: String::from("My Issues"),
                    tab_type: TabType::MyIssues,
                    custom_view: None,
                }],
            })),
        }
    }
}

#[derive(Clone, Debug, PartialEq, Default)]
enum TabType {
    #[default]
    MyIssues,
    CustomView,
    SearchResults,
}

#[derive(Debug, Clone, Default)]
struct Tab {
    title: String,
    tab_type: TabType,
    custom_view: Option<custom_views_query::ViewFragment>,
}

impl TabWidget {
    pub fn run(&self) {
        let this = self.clone();
        tokio::spawn(this.fetch());
    }

    async fn fetch(self) {
        let linear_api_token =
            std::env::var("LINEAR_API_TOKEN").expect("Missing LINEAR_API_TOKEN env var");
        let client = LinearClient::new(linear_api_token).unwrap();
        let variables = custom_views_query::Variables {};
        match client.query(CustomViewsQuery, variables).await {
            Ok(data) => {
                let mut state = self.state.write().unwrap();
                for custom_view in data.custom_views.nodes.iter() {
                    state.tabs.push(Tab {
                        title: custom_view.name.clone(),
                        tab_type: TabType::CustomView,
                        custom_view: Some(custom_view.clone()),
                    });
                }
            }
            Err(e) => {
                panic!("Error {:#?}", e);
                //return;
            }
        }
    }
    pub fn next(&self) -> usize {
        let mut state = self.state.write().unwrap();
        if state.selected_index < state.tabs.len() - 1 {
            state.selected_index += 1;
        }
        state.selected_index
    }

    pub fn prev(&self) -> usize {
        let mut state = self.state.write().unwrap();
        if state.selected_index > 0 {
            state.selected_index -= 1;
        }
        state.selected_index
    }

    pub fn show_and_select_search_tab(&self) {
        let mut state = self.state.write().unwrap();
        if state.tabs[state.tabs.len() - 1].tab_type != TabType::SearchResults {
            state.tabs.push(Tab {
                title: String::from("Search Results"),
                tab_type: TabType::SearchResults,
                custom_view: None,
            });
        }
        state.selected_index = state.tabs.len() - 1;
    }

    pub fn handle_event(&self, event: &Event) -> crate::TabChangeEvent {
        if let Event::Key(key) = event {
            if key.kind == KeyEventKind::Press {
                match key.code {
                    KeyCode::Tab => {
                        let index = self.next();
                        let state = self.state.read().unwrap();
                        match &state.tabs[index].custom_view {
                            Some(custom_view) => {
                                return TabChangeEvent::FetchCustomViewIssues(custom_view.clone());
                            }
                            _ => {
                                return match state.tabs[index].tab_type {
                                    TabType::MyIssues => TabChangeEvent::FetchMyIssues,
                                    TabType::SearchResults => TabChangeEvent::SearchIssues,
                                    _ => TabChangeEvent::None,
                                };
                            }
                        }
                    }
                    KeyCode::BackTab => {
                        self.prev();
                        let state = self.state.read().unwrap();
                        match &state.tabs[state.selected_index].custom_view {
                            Some(custom_view) => {
                                return TabChangeEvent::FetchCustomViewIssues(custom_view.clone());
                            }
                            _ => {
                                return match state.tabs[state.selected_index].tab_type {
                                    TabType::MyIssues => TabChangeEvent::FetchMyIssues,
                                    TabType::SearchResults => TabChangeEvent::SearchIssues,
                                    _ => TabChangeEvent::None,
                                };
                            }
                        }
                    }
                    _ => return TabChangeEvent::None,
                }
            }
        }
        TabChangeEvent::None
    }
}

impl Widget for &TabWidget {
    fn render(self, area: Rect, buf: &mut Buffer) {
        use Constraint::{Length, Min};
        let horizontal = Layout::horizontal([Length(23), Min(0)]);
        let [header, main] = horizontal.areas(area);

        Line::from(vec![
            Span::from(" <tab> ").blue(),
            Span::from("to change view:  "),
        ])
        .render(header, buf);
        Tabs::new(
            self.state
                .read()
                .unwrap()
                .tabs
                .iter()
                .map(|tab| {
                    let (icon, color) = if let Some(view) = &tab.custom_view {
                        (
                            match &view.icon {
                                Some(icon) => iconmap::ico_to_nf(icon),
                                _ => "".to_string(),
                            },
                            match &view.color {
                                Some(color) => String::from(color),
                                _ => "#ffffff".to_string(),
                            },
                        )
                    } else if tab.tab_type == TabType::SearchResults {
                        (iconmap::ico_to_nf("Magnify"), String::from("#FFFFFF"))
                    } else {
                        (iconmap::ico_to_nf("Home"), String::from("#FFFFFF"))
                    };
                    let project_color = Color::from_str(&color).unwrap();
                    Span::from(format!("{} {}", icon, tab.title.clone().bold()))
                        .fg(project_color)
                        .bold()
                })
                .collect::<Vec<Span>>(),
        )
        .select(self.state.read().unwrap().selected_index)
        .padding(" ", " ")
        .divider("  ")
        .render(main, buf);
    }
}

#[cfg(test)]
mod tests {
    use std::sync::{Arc, RwLock};

    use crossterm::event::{KeyCode, KeyEventKind, KeyEventState, KeyModifiers};
    use insta::assert_snapshot;
    use ratatui::{Terminal, backend::TestBackend};

    use crate::{
        TabChangeEvent,
        queries::custom_views_query,
        widgets::{TabWidget, tab_widget::TabType},
    };

    use super::{Tab, TabWidgetState};

    fn create_key_event(code: KeyCode) -> crossterm::event::Event {
        crossterm::event::Event::Key(crossterm::event::KeyEvent {
            code,
            kind: KeyEventKind::Press,
            modifiers: KeyModifiers::empty(),
            state: KeyEventState::empty(),
        })
    }

    #[test]
    fn test_empty_state() {
        let app = TabWidget::default();
        let mut terminal = Terminal::new(TestBackend::new(75, 2)).unwrap();
        terminal
            .draw(|frame| frame.render_widget(&app, frame.area()))
            .unwrap();

        assert_snapshot!(terminal.backend());

        let ev = app.handle_event(&create_key_event(KeyCode::Tab));
        assert_eq!(ev, TabChangeEvent::FetchMyIssues);

        let ev = app.handle_event(&create_key_event(KeyCode::Char('j')));
        assert_eq!(ev, TabChangeEvent::None);
    }

    #[test]
    fn test_multi_tabs() {
        let app = TabWidget {
            state: Arc::new(RwLock::new(TabWidgetState {
                selected_index: 0,
                tabs: vec![
                    Tab {
                        title: String::from("My Issues"),
                        tab_type: TabType::MyIssues,
                        custom_view: None,
                    },
                    Tab {
                        title: String::from("Custom A"),
                        tab_type: TabType::CustomView,
                        custom_view: Some(custom_views_query::ViewFragment {
                            slug_id: Some("sluga".into()),
                            color: Some("#fa0faf".to_string()),
                            icon: Some("Education".to_string()),
                            id: "sluga".into(),
                            name: "Custom A".into(),
                        }),
                    },
                    Tab {
                        title: String::from("Custom B"),
                        tab_type: TabType::CustomView,
                        custom_view: Some(custom_views_query::ViewFragment {
                            slug_id: Some("slugb".into()),
                            id: "slugb".into(),
                            color: Some("#fa0faf".to_string()),
                            icon: Some("Education".to_string()),
                            name: "Custom B".into(),
                        }),
                    },
                ],
            })),
        };

        let mut terminal = Terminal::new(TestBackend::new(75, 2)).unwrap();
        terminal
            .draw(|frame| frame.render_widget(&app, frame.area()))
            .unwrap();

        assert_snapshot!(terminal.backend());

        // should go to  tab a
        let ev = app.handle_event(&create_key_event(KeyCode::Tab));
        assert_eq!(
            ev,
            TabChangeEvent::FetchCustomViewIssues(custom_views_query::ViewFragment {
                name: "sluga".into(),
                slug_id: Some("sluga".into()),
                color: Some("#fa0faf".to_string()),
                icon: Some("Education".to_string()),
                id: String::from("sluga")
            })
        );

        // then to b
        let ev = app.handle_event(&create_key_event(KeyCode::Tab));
        assert_eq!(
            ev,
            TabChangeEvent::FetchCustomViewIssues(custom_views_query::ViewFragment {
                name: "slugb".into(),
                slug_id: Some("slugb".into()),
                color: Some("#fa0faf".to_string()),
                icon: Some("Education".to_string()),
                id: String::from("slugb")
            })
        );

        // but not passed b
        let ev = app.handle_event(&create_key_event(KeyCode::Tab));
        assert_eq!(
            ev,
            TabChangeEvent::FetchCustomViewIssues(custom_views_query::ViewFragment {
                name: "slugb".into(),
                slug_id: Some("slugb".into()),
                color: Some("#fa0faf".to_string()),
                icon: Some("Education".to_string()),
                id: String::from("slugb")
            })
        );

        // then back to a
        let ev = app.handle_event(&create_key_event(KeyCode::BackTab));
        assert_eq!(
            ev,
            TabChangeEvent::FetchCustomViewIssues(custom_views_query::ViewFragment {
                name: "sluga".into(),
                slug_id: Some("sluga".into()),
                color: Some("#fa0faf".to_string()),
                icon: Some("Education".to_string()),
                id: String::from("sluga")
            })
        );

        let ev = app.handle_event(&create_key_event(KeyCode::BackTab));
        assert_eq!(ev, TabChangeEvent::FetchMyIssues);
    }
}
