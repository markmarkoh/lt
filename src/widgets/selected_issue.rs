use crate::IssueFragment;
use crate::LtEvent;
use crate::iconmap;

use ratatui::style::Modifier;
use ratatui::style::Style;
use ratatui::style::palette::tailwind::SLATE;
use ratatui::text::Text;
use ratatui::widgets::Scrollbar;
use ratatui::widgets::ScrollbarOrientation;
use ratatui::widgets::ScrollbarState;
use ratatui::widgets::StatefulWidget;
use ratatui::widgets::Wrap;
use std::str::FromStr;
use std::sync::{Arc, RwLock};

use chrono::DateTime;

use crossterm::event::{Event, KeyCode, KeyEventKind};
use ratatui::buffer::Buffer;
use ratatui::{
    layout::{Constraint, Layout, Rect},
    style::{Color, Stylize},
    symbols,
    text::{Line, Span},
    widgets::{Block, Borders, Paragraph, Widget},
};

#[derive(Debug, Default)]
struct SelectedIssueWidgetState {
    selected_issue: Option<IssueFragment>,
}

#[derive(Debug, Clone, Default)]
pub struct SelectedIssueWidget {
    state: Arc<RwLock<SelectedIssueWidgetState>>,
    scroll_state: ScrollbarState,
    scroll: usize,
}

impl SelectedIssueWidget {
    pub fn set_selected_issue(&mut self, issue: Option<IssueFragment>) {
        self.state.write().unwrap().selected_issue = issue;
        self.scroll = 0;
        self.scroll_state = ScrollbarState::default();
    }

    pub fn handle_event(&mut self, event: &Event) -> LtEvent {
        if let Event::Key(key) = event {
            if key.kind == KeyEventKind::Press {
                match key.code {
                    KeyCode::Down => {
                        self.scroll = self.scroll.saturating_add(1);
                        self.scroll_state.next();
                    }
                    KeyCode::Up => {
                        self.scroll = self.scroll.saturating_sub(1);
                        self.scroll_state.prev();
                    }
                    _ => {}
                }
            }
        }
        LtEvent::None
    }
}

const DICT_HEADER: Style = Style::new();

fn header(text: &str) -> Line {
    Line::from(Span::from(text.to_owned() + ":\n")).style(DICT_HEADER)
}

impl Widget for &SelectedIssueWidget {
    fn render(self, area: Rect, buf: &mut Buffer) {
        let state = self.state.read().unwrap();

        let [main, scroll, sidebar] = Layout::horizontal([
            Constraint::Percentage(75),
            Constraint::Min(1),
            Constraint::Percentage(25),
        ])
        .areas(area);

        let (identifier, title_text, description, created_at, sidebar_lines) =
            match &state.selected_issue {
                Some(issue) => {
                    let identifier = issue.identifier.clone().blue().bold();
                    let title_text = Line::from(issue.title.clone()).centered();
                    let description = if issue.description.is_some() {
                        tui_markdown::from_str(issue.description.as_ref().unwrap())
                    } else {
                        tui_markdown::from_str("#### No description")
                    };
                    let created_at = DateTime::parse_from_rfc3339(&issue.created_at.clone())
                        .unwrap()
                        .format("%m/%d/%Y")
                        .to_string();

                    let mut sidebar_items = vec![];

                    if !issue.priority_label.is_empty() {
                        sidebar_items.push(header("Priority"));
                        let priority_icon = iconmap::p_to_nf(issue.priority);
                        sidebar_items.push(Line::from(vec![
                            priority_icon.add_modifier(Modifier::BOLD),
                            issue.priority_label.clone().into(),
                        ]));
                        sidebar_items.push(Line::from(""));
                    }
                    if !issue.state.name.is_empty() {
                        sidebar_items.push(header("Status"));
                        let state_icon = iconmap::state_to_nf(&issue.state.type_)
                            .fg(Color::from_str(&issue.state.color).unwrap());
                        sidebar_items.push(Line::from(vec![
                            state_icon.fg(Color::from_str(&issue.state.color).unwrap()),
                            " ".into(),
                            issue
                                .state
                                .name
                                .clone()
                                .fg(Color::from_str(&issue.state.color).unwrap()),
                        ]));
                        sidebar_items.push(Line::from(""));
                    }

                    if let Some(project) = &issue.project {
                        let project_color = Color::from_str(&project.color).unwrap();
                        let project_icon =
                            iconmap::ico_to_nf(project.icon.as_ref().map_or("NN", |v| v))
                                .fg(project_color);
                        sidebar_items.push(header("Project"));
                        sidebar_items.push(Line::from(vec![
                            project_icon,
                            project.name.clone().fg(project_color),
                        ]));
                        sidebar_items.push(Line::from(""));
                    }

                    if let Some(assignee) = &issue.assignee {
                        sidebar_items.push(header("Assignee"));
                        if assignee.is_me {
                            sidebar_items.push(Line::from("You"));
                        } else {
                            sidebar_items.push(Line::from(assignee.display_name.clone()));
                        }
                        sidebar_items.push(Line::from(""));
                    }

                    if let Some(creator) = &issue.creator {
                        sidebar_items.push(header("Creator"));
                        if creator.is_me {
                            sidebar_items.push(Line::from("You"));
                        } else {
                            sidebar_items.push(Line::from(creator.display_name.clone()));
                        }
                        sidebar_items.push(Line::from(""));
                    }

                    if !issue.labels.edges.is_empty() {
                        sidebar_items.push(header("Tags"));
                        sidebar_items.push(Line::from(
                            issue
                                .labels
                                .edges
                                .iter()
                                .map(|value| {
                                    Span::from(format!("• {} ", value.node.name.clone()))
                                        .fg(Color::from_str(&value.node.color).unwrap())
                                })
                                .collect::<Vec<Span>>(),
                        ));
                        sidebar_items.push(Line::from(""));
                    }
                    (
                        identifier,
                        title_text,
                        description,
                        created_at,
                        sidebar_items,
                    )
                }
                None => (
                    String::from("").into(),
                    Line::from(String::from("Select an issue to see some details")),
                    Text::from(""),
                    String::from(""),
                    vec![],
                ),
            };

        let created_at_title = Line::from(created_at).right_aligned();

        // collapse borders for nicer UI
        let collapsed_top_and_left_border_set = symbols::border::Set {
            top_left: symbols::line::NORMAL.horizontal_down,
            bottom_left: symbols::line::NORMAL.horizontal_up,
            ..symbols::border::PLAIN
        };

        let (effective_scroll, scroll_enabled) = if description.height() > buf.area().height.into()
        {
            let effective_scroll = if self.scroll > description.height() / 2 {
                description.height() / 2
            } else {
                self.scroll
            };

            let mut scroll_state = self
                .scroll_state
                .content_length(description.height() / 2)
                .position(effective_scroll);

            let sb = Scrollbar::new(ScrollbarOrientation::VerticalRight)
                .begin_symbol(Some("↑"))
                .end_symbol(Some("↓"));

            sb.render(scroll, buf, &mut scroll_state);
            (effective_scroll, true)
        } else {
            let placeholder_block = Block::new().borders(Borders::TOP | Borders::BOTTOM);
            placeholder_block.render(scroll, buf);
            (0, false)
        };

        let mut block = Block::new()
            .borders(Borders::TOP | Borders::LEFT | Borders::BOTTOM)
            .border_style(DICT_HEADER)
            .title_bottom(Line::from(vec![
                Span::from("──"),
                Span::from(" <y> ").blue(),
                Span::from("to yank git branch "),
            ]))
            .title_bottom(Line::from(vec![
                Span::from(" <o> ").blue(),
                Span::from("to open in Linear"),
            ]))
            .title(identifier)
            .title(title_text);

        block = if scroll_enabled {
            block.title_bottom(
                Line::from(vec![
                    Span::from(" <↑/↓> ").blue(),
                    Span::from("to scroll ──────"),
                ])
                .right_aligned(),
            )
        } else {
            block
        };

        let mut p = Paragraph::new(description.clone())
            .block(block)
            .wrap(Wrap { trim: true });

        p = p.scroll((effective_scroll as u16, 0));

        let sidebar_block = Block::bordered()
            .border_set(collapsed_top_and_left_border_set)
            .title_bottom(Line::from(vec![
                Span::from("──"),
                Span::from(" <q> ").blue(),
                Span::from("to quit "),
            ]))
            .title(created_at_title);

        let sidebar_p = Paragraph::new(sidebar_lines)
            .block(sidebar_block)
            .wrap(Wrap { trim: true });

        //Clear.render(main, buf);
        p.render(main, buf);
        sidebar_p.render(sidebar, buf);
    }
}

// Now in tests module:
#[cfg(test)]
pub(crate) mod tests {
    use crate::{
        IssueFragment, IssueFragmentAssignee, IssueFragmentCreator, IssueFragmentProject,
        IssueFragmentState,
    };
    use insta::assert_snapshot;
    use ratatui::{Terminal, backend::TestBackend};

    use crate::widgets::SelectedIssueWidget;

    pub fn make_issue(title: &str, identifier: &str) -> IssueFragment {
        IssueFragment {
            priority: 1.0,
            priority_label: "Urgent".into(),
            branch_name: "test-1-branch-name".into(),
            identifier: String::from(identifier),
            title: String::from(title),
            created_at: String::from("2025-05-10T03:09:51.740Z"),
            project: Some(IssueFragmentProject {
                name: "Test Project".into(),
                icon: Some("Subgroup".into()),
                color: "#FA0FA0".into(),
            }),
            creator: Some(IssueFragmentCreator {
                is_me: true,
                display_name: "Creator Display Name".into(),
            }),
            assignee: Some(IssueFragmentAssignee {
                is_me: false,
                display_name: "Assignee Display Name".into(),
            }),
            state: IssueFragmentState {
                name: "Backlogged".into(),
                color: "#0FA0FA".into(),
                type_: "backlog".into(),
            },
            description: Some(String::from("### Title\n\nMulti\nLine **description**")),
            ..Default::default()
        }
    }

    #[test]
    fn test_empty_state() {
        let app = SelectedIssueWidget::default();
        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
        terminal
            .draw(|frame| frame.render_widget(&app, frame.area()))
            .unwrap();
        assert_snapshot!(terminal.backend());
    }

    #[test]
    fn test_basic_issue() {
        let mut app = SelectedIssueWidget::default();
        let issue = make_issue("Testing Ticket", "TEST-1");

        app.set_selected_issue(Some(issue));

        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
        terminal
            .draw(|frame| frame.render_widget(&app, frame.area()))
            .unwrap();
        assert_snapshot!(terminal.backend());
    }
}
