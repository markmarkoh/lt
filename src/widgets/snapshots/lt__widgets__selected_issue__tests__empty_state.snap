---
source: src/widgets/selected_issue.rs
expression: terminal.backend()
---
"┌─Select an issue to see some details───────────────────────┬──────────────────┐"
"│                                                           │                  │"
"│                                                           │                  │"
"│                                                           │                  │"
"│                                                           │                  │"
"│                                                           │                  │"
"│                                                           │                  │"
"│                                                           │                  │"
"│                                                           │                  │"
"│                                                           │                  │"
"│                                                           │                  │"
"│                                                           │                  │"
"│                                                           │                  │"
"│                                                           │                  │"
"│                                                           │                  │"
"│                                                           │                  │"
"│                                                           │                  │"
"│                                                           │                  │"
"│                                                           │                  │"
"└── <y> to yank git branch ─ <o> to open in Linear──────────┴── <q> to quit ───┘"
